bindings/wasm/output.go

```diff
@@ -0,0 +1,208 @@
+	moduleConfig  wazero.ModuleConfig
+
+	binaryName string
+	runtime    wazero.Runtime
+	module     wazero.CompiledModule
+
+	instanceCounter atomic.Uint64
+}
+
+var (
+	_ bindings.OutputBinding = (*outputBinding)(nil)
+	_ io.Closer              = (*outputBinding)(nil)
+)
+
+func NewWasmOutput(logger logger.Logger) bindings.OutputBinding {
+	return &outputBinding{
+		logger: logger,
+
+		// The below ensures context cancels in-flight wasm functions.
+		runtimeConfig: wazero.NewRuntimeConfig().
+			WithCloseOnContextDone(true),
+
+		// The below violate sand-boxing, but allow code to behave as expecteout.
+		moduleConfig: wazero.NewModuleConfig().
+			WithRandSource(rand.Reader).
+			WithSysNanosleep().
+			WithSysWalltime().
+			WithSysNanosleep(),
+	}
+}
+
+func (out *outputBinding) Init(ctx context.Context, metadata bindings.Metadata) (err error) {
+	meta, err := out.getInitMetadata(metadata)
+	if err != nil {
+		return fmt.Errorf("wasm: failed to parse metadata: %w", err)
+	}
+
+	// Use the name of the wasm binary as the module name.
+	out.binaryName, _ = strings.CutSuffix(path.Base(meta.Path), ".wasm")
+
+	// Create the runtime, which when closed releases any resources associated with it.
+	out.runtime = wazero.NewRuntimeWithConfig(ctx, out.runtimeConfig)
+
+	// Compile the module, which reduces execution time of Invoke
+	out.module, err = out.runtime.CompileModule(ctx, meta.guest)
+	if err != nil {
+		_ = out.runtime.Close(context.Background())
+		return fmt.Errorf("wasm: error compiling binary: %w", err)
+	}
+
+	switch detectImports(out.module.ImportedFunctions()) {
+	case modeWasiP1:
+		_, err = wasi_snapshot_preview1.Instantiate(ctx, out.runtime)
+	}
+
+	if err != nil {
+		_ = out.runtime.Close(context.Background())
+		return fmt.Errorf("wasm: error instantiating host functions: %w", err)
+	}
+	return
+}
+
+func (out *outputBinding) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// Currently, concurrent modules can conflict on name. Make sure we have
+	// a unique one.
+	instanceNum := out.instanceCounter.Add(1)
+	instanceName := out.binaryName + "-" + strconv.FormatUint(instanceNum, 10)
+	moduleConfig := out.moduleConfig.WithName(instanceName)
+
+	// Only assign STDIN if it is present in the request.
+	if len(req.Data) > 0 {
+		moduleConfig = moduleConfig.WithStdin(bytes.NewReader(req.Data))
+	}
+
+	// Any STDOUT is returned as a result: capture it into a buffer.
+	var stdout bytes.Buffer
+	moduleConfig = moduleConfig.WithStdout(&stdout)
+
+	// Set the program name to the binary name
+	argsSlice := []string{out.binaryName}
+
+	// Get any remaining args from configuration
+	if args := req.Metadata["args"]; args != "" {
+		argsSlice = append(argsSlice, strings.Split(args, ",")...)
+	}
+	moduleConfig = moduleConfig.WithArgs(argsSlice...)
+
+	// Instantiating executes the guest's main function (exported as _start).
+	_, err := out.runtime.InstantiateModule(ctx, out.module, moduleConfig)
```

Comment:
@codefromthecrypt You know best. Even though `_start` runs to completion, do you need to call `module.Close()` to clean up any resources in wazero?

Would it make sense to add `defer module.Close()`?

Reply:
good eye!

Reply:
https://github.com/dapr/components-contrib/pull/2697